### ISSUE TITLE
Incorrect problem detected by void_return rule when returning async functions

### ISSUE THREAD

**Issue by lo1tuma** [realm/SwiftLint#4766]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

The following example is detected as a problem by the `void_return` rule:

```swift
let bar: () -> () async -> Void = { return {} }
```

##### Complete output when running SwiftLint, including the stack trace and command used

```
Void Return Violation: Prefer `-> Void` over `-> ()`. (void_return)
```

### Environment

* SwiftLint version: I’m using the revision #7eb479d546a0cc099dae473cadccba303cc8a00a
* Installation method used: swift-package-manager via tuist
* Paste your configuration file:

<details>

```yml
only_rules:
    - block_based_kvo
    - class_delegate_protocol
    - compiler_protocol_init
    - cyclomatic_complexity
    - deployment_target
    - discouraged_direct_init
    - duplicate_enum_cases
    - duplicate_imports
    - duplicated_key_in_dictionary_literal
    - dynamic_inline
    - empty_enum_arguments
    - empty_parameters
    - empty_parentheses_with_trailing_closure
    - for_where
    - force_cast
    - force_try
    - function_body_length
    - function_parameter_count
    - identifier_name
    - is_disjoint
    - large_tuple
    - legacy_cggeometry_functions
    - legacy_constant
    - legacy_constructor
    - legacy_hashing
    - legacy_nsgeometry_functions
    - legacy_random
    - multiple_closures_with_trailing_closure
    - nesting
    - no_fallthrough_only
    - notification_center_detachment
    - ns_number_init_as_function_reference
    - nsobject_prefer_isequal
    - private_over_fileprivate
    - private_unit_test
    - reduce_boolean
    - redundant_discardable_let
    - redundant_objc_attribute
    - redundant_optional_initialization
    - redundant_set_access_control
    - redundant_string_enum_value
    - self_in_property_initialization
    - superfluous_disable_command
    - syntactic_sugar
    - todo
    - unavailable_condition
    - unneeded_break_in_switch
    - unused_closure_parameter
    - unused_control_flow_label
    - unused_enumerated
    - unused_optional_binding
    - unused_setter_value
    - valid_ibinspectable
    - void_function_in_ternary
    - void_return
    - xctfail_message
    - accessibility_label_for_image
    - accessibility_trait_for_button
    - anonymous_argument_in_multiline_closure
    - anyobject_protocol
    - array_init
    - closure_body_length
    - contains_over_filter_count
    - contains_over_filter_is_empty
    - contains_over_first_not_nil
    - contains_over_range_nil_comparison
    - convenience_type
    - discarded_notification_center_observer
    - discouraged_assert
    - discouraged_none_name
    - discouraged_object_literal
    - discouraged_optional_boolean
    - discouraged_optional_collection
    - empty_collection_literal
    - empty_count
    - empty_string
    - empty_xctest_method
    - enum_case_associated_values_count
    - explicit_init
    - fallthrough
    - fatal_error_message
    - file_header
    - file_name_no_space
    - first_where
    - flatmap_over_map_reduce
    - force_unwrapping
    - identical_operands
    - implicitly_unwrapped_optional
    - inert_defer
    - joined_default_parameter
    - last_where
    - legacy_multiple
    - lower_acl_than_parent
    - no_extension_access_modifier
    - no_grouping_extension
    - number_separator
    - optional_enum_case_matching
    - override_in_extension
    - pattern_matching_keywords
    - prefer_self_in_static_references
    - prefer_self_type_over_type_of_self
    - prefer_zero_over_explicit_init
    - private_action
    - private_outlet
    - private_subject
    - quick_discouraged_call
    - quick_discouraged_focused_test
    - quick_discouraged_pending_test
    - redundant_nil_coalescing
    - redundant_type_annotation
    - return_value_from_void_function
    - self_binding
    - shorthand_optional_binding
    - single_test_class
    - sorted_first_last
    - strict_fileprivate
    - strong_iboutlet
    - test_case_accessibility
    - toggle_bool
    - unavailable_function
    - untyped_error_in_catch
    - unused_capture_list
    - yoda_condition

analyzer_rules:
    - capture_variable
    - unused_declaration
    - unused_import
    - typesafe_array_init
    - explicit_self


block_based_kvo: error

class_delegate_protocol: error

compiler_protocol_init: error

cyclomatic_complexity:
    warning: 10
    error: 10
    ignores_case_statements: false

deployment_target:
    severity: error
    iOS_deployment_target: '13'

discouraged_direct_init:
    severity: error

duplicate_enum_cases: error

duplicate_imports: error

duplicated_key_in_dictionary_literal: error

dynamic_inline: error

empty_enum_arguments: error

empty_parameters: error

empty_parentheses_with_trailing_closure: error

for_where:
    severity: error
    allow_for_as_filter: false

force_cast: error

force_try: error

function_body_length:
    warning: 50
    error: 50

function_parameter_count:
    warning: 5
    error: 5
    ignores_default_parameters: true

identifier_name:
    min_length:
        warning: 2
        error: 2
    max_length:
        warning: 80
        error: 80
    allowed_symbols: "_"
    excluded:
        - x
        - y
        - z
        - w

is_disjoint: error

large_tuple:
    warning: 3
    error: 3

legacy_cggeometry_functions: error

legacy_constant: error

legacy_constructor: error

legacy_hashing: error

legacy_nsgeometry_functions: error

legacy_random: error

multiple_closures_with_trailing_closure: error

nesting:
    severity: error
    type_level: 1
    function_level: 2
    check_nesting_in_closures_and_statements: true
    always_allow_one_type_in_functions: false

no_fallthrough_only: error

notification_center_detachment: error

ns_number_init_as_function_reference: error

nsobject_prefer_isequal: error

private_over_fileprivate:
    severity: error
    validate_extensions: false

private_unit_test:
    severity: error

reduce_boolean: error

redundant_discardable_let: error

redundant_objc_attribute: error

redundant_optional_initialization: error

redundant_set_access_control: error

redundant_string_enum_value: error

self_in_property_initialization: error

superfluous_disable_command: error

syntactic_sugar: error

todo: error

unavailable_condition: error

unneeded_break_in_switch: error

unused_closure_parameter: error

unused_control_flow_label: error

unused_enumerated: error

unused_optional_binding:
    severity: error
    ignore_optional_try: false

unused_setter_value: error

valid_ibinspectable: error

void_function_in_ternary: error

void_return: error

xctfail_message: error

accessibility_label_for_image: error

accessibility_trait_for_button: error

anonymous_argument_in_multiline_closure: error

anyobject_protocol: error

array_init: error

capture_variable: error

closure_body_length:
    warning: 50
    error: 50

contains_over_filter_count: error

contains_over_filter_is_empty: error

contains_over_first_not_nil: error

contains_over_range_nil_comparison: error

convenience_type: error

discarded_notification_center_observer: error

discouraged_assert: error

discouraged_none_name: error

discouraged_object_literal:
    severity: error

discouraged_optional_boolean: error

discouraged_optional_collection: error

empty_collection_literal: error

empty_count:
    severity: error
    only_after_dot: false

empty_string: error

empty_xctest_method:
    severity: error

enum_case_associated_values_count:
    warning: 5
    error: 5

explicit_init: error

explicit_self: error

fallthrough: error

fatal_error_message: error

file_header:
    severity: error

file_name_no_space:
    severity: error

first_where: error

flatmap_over_map_reduce: error

force_unwrapping: error

identical_operands: error

implicitly_unwrapped_optional:
    severity: error
    mode: all_except_iboutlets

inert_defer: error

joined_default_parameter: error

last_where: error

legacy_multiple: error

legacy_objc_type: error

lower_acl_than_parent: error

no_extension_access_modifier: error

no_grouping_extension: error

no_magic_numbers: error

number_separator:
    severity: error
    minimum_length: 4

optional_enum_case_matching: error

override_in_extension: error

pattern_matching_keywords: error

prefer_self_in_static_references: error

prefer_self_type_over_type_of_self: error

prefer_zero_over_explicit_init: error

private_action: error

private_outlet:
    severity: error
    allow_private_set: false

private_subject: error

prohibited_interface_builder: error

quick_discouraged_call: error

quick_discouraged_focused_test: error

quick_discouraged_pending_test: error

redundant_nil_coalescing: error

redundant_type_annotation: error

return_value_from_void_function: error

self_binding:
    severity: error

shorthand_optional_binding: error

single_test_class:
    severity: error

sorted_first_last: error

strict_fileprivate: error

strong_iboutlet: error

test_case_accessibility:
    severity: error

toggle_bool: error

typesafe_array_init: error

unavailable_function: error

untyped_error_in_catch: error

unused_capture_list: error

unused_declaration:
    severity: error

unused_import:
    severity: error

yoda_condition: error
```

</details>

* Which Xcode version are you using? 14.2
